metrics/registry.go

```diff
@@ -0,0 +1,138 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+func NewNamespaceRegistry(n string) *NamespaceRegistry {
+	return &NamespaceRegistry{
+		Namespace: n,
+		mr:        NewMetricDescRegistry(),
+		m:         &sync.Map{},
+	}
+}
+
+type NamespaceRegistry struct {
+	Namespace string
+
+	mr *MetricDescRegistry
+
+	// Map of collectable -> *MetricDescRegistry
+	m *sync.Map
+}
+
+func (n *NamespaceRegistry) Describe(c chan<- MetricDesc) error {
+	c <- MetricDesc{
+		Name:   n.Namespace,
+		Prefix: true,
+	}
+	return nil
+}
+
+// Collect simply calls collect on all the collectables in this registry adding its
+// namespace as a prefix to the name
+func (n *NamespaceRegistry) Collect(ctx context.Context, points chan<- MetricPoint) error {
+	f := func(c Collectable, r *MetricDescRegistry) error {
+		var err error
+		// We need to call collect on the children and add our namespace stuff to the value that is returned
+		innerPoints := make(chan MetricPoint)
+		go func() {
+			defer close(innerPoints)
+			err = c.Collect(ctx, innerPoints)
+		}()
+
+	WAITRESULT:
+		for {
+			select {
+			case item, ok := <-innerPoints:
+				if !ok {
+					break WAITRESULT
+				}
+				if !r.Contains(item.Name) {
+					fmt.Printf("Skipping item %s as it wasn't present for Describe() (not in %v): %v", item.Name, r.prefixTree.ToMap(), item)
+					continue
+				}
+				if n.Namespace != "" {
+					item.Metric.Name = n.Namespace + "_" + item.Metric.Name
+				}
+				points <- item
+			}
+		}
+		return err
+	}
+
+	return n.Each(ctx, f)
+}
+
+func (n *NamespaceRegistry) Register(c Collectable) error {
+	var err error
+	ch := make(chan MetricDesc)
+	go func() {
+		defer close(ch)
+		err = c.Describe(ch)
+	}()
+
+	metricDescs := make([]MetricDesc, 0)
```

Comment:
nit: I generally use `[]MetricDesc{}` when making an empty one as I use the call to `make` to signify that something special is happening. I can't recall where I learned that approach, but I find value in being able to quickly visually determine which case it is. No need to change it here, though.